File: lib/models/post.dart
import 'user.dart';

class Post {
  final String id;
  final String content;
  final String? image;
  final User author;
  final String timestamp;
  final int likes;
  final int comments;
  final int shares;

  Post({
    required this.id,
    required this.content,
    this.image,
    required this.author,
    required this.timestamp,
    this.likes = 0,
    this.comments = 0,
    this.shares = 0,
  });

  factory Post.fromJson(Map<String, dynamic> json) {
    return Post(
      id: json['_id'] ?? json['id'] ?? '',
      content: json['content'] ?? '',
      image: json['image'] ?? json['image_url'],
      author: json['author'] != null 
          ? User.fromJson(json['author']) 
          : User(id: 'unknown', name: 'Unknown', username: 'unknown'),
      timestamp: json['timestamp'] ?? json['createdAt'] ?? DateTime.now().toIso8601String(),
      likes: (json['likes'] is List) 
          ? (json['likes'] as List).length 
          : (json['likes'] ?? 0),
      comments: json['comments'] ?? 0,
      shares: json['shares'] ?? 0,
    );
  }
}

File: backend/models/Post.js
const mongoose = require('mongoose');

const PostSchema = new mongoose.Schema({
  author: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'User', 
    required: true 
  },
  content: { type: String },
  image_url: { type: String },
  title: { type: String }, // Optional title
  
  // Interactions
  // Thay đổi từ mảng ID sang mảng Object để lưu loại reaction
  reactions: [{
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    type: { type: String, enum: ['like', 'love', 'haha', 'wow', 'sad', 'angry'], default: 'like' }
  }],
  shares: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
  
  // Comments sẽ được query từ collection Comment, nhưng có thể cache số lượng nếu cần
  
  createdAt: { type: Date, default: Date.now },
  editedAt: { type: Date }, // Thời gian chỉnh sửa
  originalPost: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Post' 
  } // Bài viết gốc nếu là chia sẻ
});

module.exports = mongoose.model('Post', PostSchema);